### ISSUE TITLE
Better Start at Login?

### ISSUE THREAD

**Issue by avaidyam** [bikkelbroeders/TouchBarDemoApp#19]
Maybe a good idea would be to just use the `LSSharedFileList` API even though it's soft-deprecated? I think the use of a separate helper app is a bit unwieldy. Also, the repo could also be renamed to `TouchBarServer` or `TouchBar` depending on what the macOS app's name will end up being.

**Comment by robbertkl** [comment]
I thought this is (or was?) the recommended way to do the Start on Login. I also don't really like the helper app, but it works, as long as the app resides in `/Applications` anyway 😉. Feel free to drop a PR to change it and I'll take a look. Please include the pros and cons for the other method so we can use that in our decision.

As for the repo rename, what do you mean exactly? The repo currently contains both client and server, so naming it TouchBarServer seems wrong. We chose to add "DemoApp" to make it clear to everyone (and to Apple) that it's just meant as a demo.

**Comment by avaidyam** [comment]
As for the repo renaming, I know I use this tool the same as the Xcode built in simulator for development purposes... so why not call it `TouchBarSimulator`? 

> To create a login item for your sandboxed app, use the SMLoginItemSetEnabled function (declared in ServiceManagement/SMLoginItem.h) as described in “Adding Login Items Using the Service Management Framework” in Daemons and Services Programming Guide.
> (With App Sandbox, you cannot create a login item using functions in the LSSharedFileList.h header file. For example, you cannot use the function LSSharedFileListInsertItemURL. Nor can you manipulate the state of launch services, such as by using the function LSRegisterURL.)

This tool is probably never going on the MAS so we should be able to use the _much cleaner_ API.

**Comment by robbertkl** [comment]
We'll consider a repo name change, but now that it's still getting linked a lot I think we'll leave it until it cools down a bit.

As for the _much cleaner_ API: I'm all for it! You're absolutely right by saying this will probably never go on the MAS. I'll keep this issue open so we can check it out later. In the meantime, feel free to send a PR if you can't wait :wink:.

**Comment by w0lfschild** [comment]
Renamed repos will redirect from it's old URL. Not sure if that lasts forever but TouchBarDemoApp is an odd repo name. https://github.com/blog/1508-repository-redirects-are-here

You can use applescript. To add or remove something from the users login items. Not the most elegant solution but it works.

``` objectivec
- (IBAction)startAtLogin:(id)sender {
    if ([sender state] == NSOffState)
    {
        dispatch_queue_t myQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
        dispatch_async(myQueue, ^{
            NSString *appPath = [[NSBundle mainBundle] bundlePath];
            NSString *applescript =  [NSString stringWithFormat:@"\
                                      tell application \"System Events\"\n\
                                      if (login item \"TouchBarServer\" exists) is false then\n\
                                      make new login item at end of login items with properties {path:\"%@\", hidden:false}\n\
                                      else\n\
                                      delete login item \"TouchBarServer\"\n\
                                      make new login item at end of login items with properties {path:\"%@\", hidden:false}\n\
                                      end if\n\
                                      end tell", appPath, appPath];
            NSDictionary* errorDict;
            NSAppleEventDescriptor *returnDescriptor;
            NSAppleScript* scriptObject = [[NSAppleScript alloc] initWithSource:applescript];
            returnDescriptor = [scriptObject executeAndReturnError: &errorDict];
        });
    } else {
        dispatch_queue_t myQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
        dispatch_async(myQueue, ^{
            NSString *applescript =  [NSString stringWithFormat:@"\
                                      tell application \"System Events\"\n\
                                      if (login item \"TouchBarServer\" exists) is true then\n\
                                      delete login item \"TouchBarServer\"\n\
                                      end if\n\
                                      end tell"];
            NSDictionary* errorDict;
            NSAppleEventDescriptor *returnDescriptor;
            NSAppleScript* scriptObject = [[NSAppleScript alloc] initWithSource:applescript];
            returnDescriptor = [scriptObject executeAndReturnError: &errorDict];
        });
    }
    [sender setState:![sender state]];
}
```

**Comment by sindresorhus** [comment]
You might find [`LaunchAtLogin`](https://github.com/sindresorhus/LaunchAtLogin) useful. It only requires a couple of lines of code to use and no need for you to deal with a helper app.

**Comment by robbertkl** [comment]
Yes, that was already mentioned in #14, which was closed in favour of this one. PR's for `LaunchAtLogin` are welcome!

**Comment by robbertkl** [comment]
@sindresorhus This uses a helper app as well and creates an external dependency, so rather not.

PR's for `LSSharedFileList` stuff are welcome, though.
